### ISSUE TITLE
TransportError: ('Connection aborted.', BadStatusLine("''",))

### ISSUE THREAD

**Issue by BRIDGE-AI** [GoogleCloudPlatform/python-docs-samples#2377]
## In which file did you encounter the issue?

I encounter timeout error while making request to `https://oauth2.googleapis.com/token` for gcp auth.

I found `socket.getaddrinfo()` in `urllib3/util/connection.py`(line 57 below) has blocked while 20 seconds and makes timeout error when the app calls `gcp_ttsclient.list_voices()`.

```python
 # /usr/lib/python2.7/site-packages/urllib3/util/connection.py
 33 def create_connection(address, timeout=socket._GLOBAL_DEFAULT_TIMEOUT,
 34                       source_address=None, socket_options=None):
 ...
 56
 57     for res in socket.getaddrinfo(host, port, family, socket.SOCK_STREAM):
 58         af, socktype, proto, canonname, sa = res
 59         sock = None
 60         try:
 61             sock = socket.socket(af, socktype, proto)
 62
 63             # If provided, set socket level options before connecting.
 64             _set_socket_options(sock, socket_options)
 65
 66             if timeout is not socket._GLOBAL_DEFAULT_TIMEOUT:
 67                 sock.settimeout(timeout)
 68             if source_address:
 69                 sock.bind(source_address)
 70             sock.connect(sa)
 71             return sock
 72
 73         except socket.error as e:
 74             err = e
 75             if sock is not None:
 76                 sock.close()
 77                 sock = None
 78
 79     if err is not None:
 80         raise err
 81
 82     raise socket.error("getaddrinfo returns an empty list")
```

I've got the error on below code.

```python
def _init_gcp2(gcp_credentials):
    from google.cloud import texttospeech
    from google.cloud.texttospeech import enums
    from google.oauth2 import service_account
    import json

    global gcp_ttsclient

    if gcp_ttsclient:
        return

    gcp_credentials = { "type": "service_account", "project_id": ...}
    creds = service_account.Credentials.from_service_account_info(gcp_credentials)
    gcp_ttsclient = texttospeech.TextToSpeechClient(credentials=creds)

    # Performs the list voices request
    voices = gcp_ttsclient.list_voices()

    for voice in voices.voices:
        id     = voice.name
        name   = "-".join(voice.name.split("-")[2:])
        ...

    ...
```

- error message
```
urllib3.connectionpool 2019-09-02 13:43:56.910 [DEBUG] Starting new HTTPS connection (2): oauth2.googleapis.com:443                                                             (connectionpool.py:813)
google.api_core.retry 2019-09-02 13:44:16.907 [DEBUG] Retrying due to 503 Deadline Exceeded, sleeping 0.2s ...                                                                 (retry.py:206)
grpc._plugin_wrapping 2019-09-02 13:44:16.914 [ERROR] AuthMetadataPluginCallback "<google.auth.transport.grpc.AuthMetadataPlugin object at 0x7f70080ba2d0>" raised exception!  (_plugin_wrapping.py:83)
Traceback (most recent call last):
  File "/usr/lib64/python2.7/site-packages/grpc/_plugin_wrapping.py", line 79, in __call__
    callback_state, callback))
  File "/usr/lib/python2.7/site-packages/google/auth/transport/grpc.py", line 77, in __call__
    callback(self._get_authorization_headers(context), None)
  File "/usr/lib/python2.7/site-packages/google/auth/transport/grpc.py", line 65, in _get_authorization_headers
    headers)
  File "/usr/lib/python2.7/site-packages/google/auth/credentials.py", line 122, in before_request
    self.refresh(request)
  File "/usr/lib/python2.7/site-packages/google/oauth2/service_account.py", line 322, in refresh
    request, self._token_uri, assertion)
  File "/usr/lib/python2.7/site-packages/google/oauth2/_client.py", line 145, in jwt_grant
    response_data = _token_endpoint_request(request, token_uri, body)
  File "/usr/lib/python2.7/site-packages/google/oauth2/_client.py", line 106, in _token_endpoint_request
    method='POST', url=token_uri, headers=headers, body=body)
  File "/usr/lib/python2.7/site-packages/google/auth/transport/requests.py", line 298, in __call__
    six.raise_from(new_exc, caught_exc)
  File "/usr/lib64/python2.7/site-packages/six.py", line 737, in raise_from
    raise value
TransportError: ('Connection aborted.', BadStatusLine("''",))
```

### Did you change the file? If so, how?

No.

## Describe the issue

I'm integrating cloud testtospeech on my server.

gcp_ttsclient tries to reconnect 2 or 3 times more and connected finally.
Then I can synthesis text to speech well.
But it's wasting time for 40~60 seconds on initializing.

I heard that 'socket.getaddrinfo()' is very primitive function and it works quietly well.
I hope this issue could be resolved by simple modification of config.
But I cannot find where I start to analyze and resolve this issue for.

I'm working on amazon linux and flask+gunicorn environment under python 2.7.14.
And I'm using google-cloud-texttospeech==0.5.0.

Please help me.

**Comment by nnegrey** [comment]
Hi, to better understand your set up. 

Are you doing this kind of flow?
1. Call https://oauth2.googleapis.com/token to get a token
2. Create GCP credentials from token
3. Pass credentials to your `_init_gcp2` method
4. Calling Text-to-Speech

Is it possible for you to use a Service Account on your server so that you don't have to call out to https://oauth2.googleapis.com/token for a token?

**Comment by BRIDGE-AI** [comment]
Hi, @nnegrey 

It's a little different.

I'm doing like this.

1. Load credential file that I downloaded from gcp console(related with my Service Account)
  `gcp_credentials = { "type": "service_account", "project_id": ...}`
  passing 'gcp_credentials' to `_init_gcp2` is not meaningful now

2. Create GCP credentials
  `creds = service_account.Credentials.from_service_account_info(info)`

3. Call Text-to-Speech
  `voices = gcp_ttsclient.list_voices()`

I thought I'm already using Service Account.
And then I don't have to call out to https://oauth2.googleapis.com/token.
But gcp_ttsclient.list_voices() is calling out `/token` now.

I'd like to use Service Account string instead the file formatted json.
You may teach me how I create GCP credentials from it correctly.

**Comment by nnegrey** [comment]
Ah. Gotcha, I think you actually need to use `from_service_account_file` instead of `from_service_account_info`. I think that should do the trick.

**Comment by BRIDGE-AI** [comment]
@nnegrey 

`from_service_account_info` may not be a parallel interface for object against `from_service_account_info` for `.json` file.
The name is making me confused enough.

Actually I don't want to use individual .json file.
I'd like to keep the json string in my config file instead the .json file.
And then I can unify all the configurations related with my project as only a file.

So, is there any other interface which fit to me?

**Comment by nnegrey** [comment]
Check out this page: https://google-auth.readthedocs.io/en/latest/reference/google.oauth2.service_account.html

Might have more info on your specific use case with the oauth2 service.

**Comment by BRIDGE-AI** [comment]
According to the reference,
`google.oauth2.service_account` has 2 interfaces `from_service_account_file` and `from_service_account_info`. [[class Credentials](https://google-auth.readthedocs.io/en/latest/reference/google.oauth2.service_account.html#google.oauth2.service_account.Credentials)]

> Or if you already have the service account file loaded:\
\
\`\`\`\
service_account_info = json.load(open('service_account.json'))\
credentials = service_account.Credentials.from_service_account_info(\
    service_account_info)\
\`\`\`

It may mean that `from_service_account_info` is a parallel interface for object against `from_service_account_info`.
I can confirm this by comparing the source of these 2 methods. [[source](https://google-auth.readthedocs.io/en/latest/_modules/google/oauth2/service_account.html#Credentials.from_service_account_info)]

This process 'https://github.com/GoogleCloudPlatform/python-docs-samples/issues/2377#issuecomment-527811792' and the use case might not be wrong.

But I cannot recognize why this problem issues.
I think I need more analysis about this.

**Comment by BRIDGE-AI** [comment]
I tried several things.

1. Using service-agent.json
I rewrote `_init_gcp2()` which takes json file name, json string or object for gcp service-account like below.

```
def _init_gcp2(credinfo):
    # credinfo : 'file name', 'json string' or 'object' for gcp service-account

    from google.cloud import texttospeech
    from google.cloud.texttospeech import enums
    from google.oauth2 import service_account
    import json

    #credinfo = { "type": "service_account", "project_id": ...}

    credinfo_obj = None

    if isinstance(credinfo, str) or isinstance(credinfo, unicode):
        try:
            credinfo_obj = json.loads(credinfo)
        except:
            pass
    else:
        credinfo_obj = credinfo

    if credinfo_obj:
        creds = service_account.Credentials.from_service_account_info(credinfo_obj)
    else:
        creds = service_account.Credentials.from_service_account_file(credinfo)

    client = texttospeech.TextToSpeechClient(credentials=creds)

    # Performs the list voices request
    voices = client.list_voices()

    for voice in voices.voices:
        id     = voice.name
        name   = "-".join(voice.name.split("-")[2:])
        ...
```

And I passed `service-account.json` to it, but the problem is retained still.

I assume that the credential's trying to access `/token` is normal and to be expected on calling `client.list_voices()`.

And then, I cannot understand why initial call for `list_voices()` make `timeout` and how finally connected successfully after 2 or 3 times after.
For every time consistently.

In conclusion, this use case is not working for me.

2. Running `_init_gcp2()` independently without server

I'm working on flask, gunicorn and gevent coroutine environment together.
I removed all the other modules and run `_init_gcp2()` alone.
And then, it works fine in the use case that use 'json string' or '.json file'.

This problem should be analyzed related with server environment.
Do you have any idea about this?

Please help me more.
Which direction I should see and dig?

**Comment by nnegrey** [comment]
Can you run this script locally and verify that it works?

```
import json
from google.oauth2 import service_account
from google.cloud import texttospeech

info = json.load(open('path_to_your_service_account_file.json'))
credentials = service_account.Credentials.from_service_account_info(info)

client = texttospeech.TextToSpeechClient(credentials=credentials)

print(client.list_voices())
```

**Comment by BRIDGE-AI** [comment]
@nnegrey 

yes, it is working on environment locally.

**Comment by nnegrey** [comment]
The likely problem is with your server then and not the code or libraries for text-to-speech.

Does it occur when your server first starts up or does it happen later?

**Comment by BRIDGE-AI** [comment]
When the server first starts up.

If then I should be able to share my code after make its structure compact and re-occurred the problem.


Can I doubt the harmony of grpc implementation and some communication or concurrency related modules?
Or are there any problems with the monkey patch of the python coroutine module ‘gevent’?

**Comment by nnegrey** [comment]
Is this an EC2 instance?

**Comment by BRIDGE-AI** [comment]
Yes.

**Comment by nnegrey** [comment]
If you run a barebones server on your local machine, does it work then?

Just to make sure it isn't an EC2 thing.

**Comment by BRIDGE-AI** [comment]
I’ll do test and give you answer.
I need a little time for it.
Thank you.
